File: day010/203. 移除链表元素.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def removeElements(self, head: ListNode, val: int) -> ListNode:
        # 时间复杂度O(N)，空间复杂度O(1)
        if head is None:
            return head
        while head.val == val:
            head = head.next
            if head is None:
                break
        pre, cur = None, head
        while cur is not None:
            if cur.val == val:
                cur = cur.next
                pre.next = cur
                continue
            pre = cur
            cur = cur.next
        return head

File: day010/203. 移除链表元素.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* removeElements(ListNode* head, int val) {
        if (head == NULL) return head;
        while (head && head->val == val) head = head->next;
        if (head == NULL) return head;
        ListNode* p = head;
        ListNode* cur = p->next;
        while (cur)
        {
            if (cur->val == val)
            {
                p->next = cur->next;
                cur = cur->next;
            }
            else
            {
                p = cur;
                cur = cur->next;
            }
        }
        return head;
    }
};
